src/esperanza/validatorstate.h

```diff
@@ -17,56 +17,71 @@
 namespace esperanza {
 
 // clang-format off
+// All phases are ordered in a way how they can progress.
+// Every phase changes either one step up and down (if re-org happens).
```

Comment:
> Every phase changes either one step up and down (if re-org happens).

Although it's correct generally, the phrase `phase change` contradicts with proposed changes because phase is computed dynamically and depends on one of FinalizationState-s rather than single ValidatorState.

Reply:
I wanted to highlight how the phase can progress. The reader or the next one who touches this code will know how the phase supposed to change. 
Will think how to rephrase keeping the same meaning. If you have a proposal, feel free to share.

Reply:
```C++
//! \brief Represents the finalizer's phase.
//!
//! Refer `esperanza::WalletExtention` for details.
//! Shortly lifecycle of finalizer is described inline, it's supposed to progress linear:
enum {
//! Finalizer doesn't vote, it's a phase of newly created validator, or after being withdraw.
NOT_VALIDATING = 0,
//! Deposit is in the mempool/wallet but is not included in a block (could happen after `deposit` rpc).
WAITING_DEPOSIT_CONFIRMATION = 1,
//! Deposit confirmed, waiting reaching the dynasty where it is allowed to vote.
WAITING_START_DYNASTY = 2,
//! Finalizer is able to vote. Starts from Validator.m_start_dynasty
//! and until logout delay passes.
IS_VALIDATING = 3,
//! Logged out (by `logout` rpc).
WAITING_FOR_WITHDRAW_DELAY = 4,
//! Withdraw delay passed but finalizer hasn't withdrawn yet (by `withdraw` rpc).
WAITING_TO_WITHDRAW = 5,
};
```

I don't really see why we need sparse steps here actually. We don't even any numbers, actually.

Reply:
@frolosofsky `IS_VALIDATING` represents 2 phases:
VALIDATING_BEFORE_LOGOUT
VALIDATING_AFTER_LOGOUT
This is at least one thing that I have in mind which might be useful to track but at this moment I didn't find use cases so left as it is. I can expect that before release we might have more intermediate steps to provide better user experience (more granular error from the RPC).

Having sparse index makes the migration easier once we freeze the protocol (make it backward compatible)

Reply:
In your example comments, you use `(by logout rpc).`. Want to clarify on this.

Phase can change not only when RPC is issued but also when the block is processed. For example you re-start the node (block is read from disk and re-processed) or you do a full-sync again.

Reply:
It's not a part of protocol. But anyway it's nit.

Yeah, it must be `(for example by logout rpc)` or alike. I find rpc examples are most user friendly way to describe this workflow.

Reply:
I removed controvetial sentence: `Every phase changes either one step up and down (if re-org happens).`